row_buffer.go

```diff
@@ -0,0 +1,505 @@
+//go:build go1.18
+
+package parquet
+
+import (
+	"io"
+	"sort"
+
+	"github.com/segmentio/parquet-go/deprecated"
+	"github.com/segmentio/parquet-go/encoding"
+	"github.com/segmentio/parquet-go/internal/unsafecast"
+)
+
+// RowBuffer is an implementation of the RowGroup interface which stores parquet
+// rows in memory.
+//
+// Unlike GenericBuffer which uses a column layout to store values in memory
+// buffers, RowBuffer uses a row layout. The use of row layout provides greater
+// efficiency when sorting the buffer, which is the primary use case for the
+// RowBuffer type. Applications which intend to sort rows prior to writing them
+// to a parquet file will often see lower CPU utilization from using a RowBuffer
+// than a GenericBuffer.
+//
+// RowBuffer values are not safe to use concurrently from multiple goroutines.
+type RowBuffer[T any] struct {
+	schema  *Schema
+	sorting []SortingColumn
+	buffer  []byte
+	rows    []Row
+	numRows int
+	compare func(Row, Row) int
+}
+
+// NewRowBuffer constructs a new row buffer.
+func NewRowBuffer[T any](options ...RowGroupOption) *RowBuffer[T] {
+	config := DefaultRowGroupConfig()
+	config.Apply(options...)
+	if err := config.Validate(); err != nil {
+		panic(err)
```

Comment:
Generally I don't expect a  function called `New` to panic unless the options are _really_ obviously wrong. But I guess we should document that if so.

Reply:
This is the pattern we've used in this library, the application can validate the configuration prior to calling the function if it needs to validate it.

Passing invalid configuration seems like something obviously wrong to me, but that's a personal appreciation obviously :)